share/man/man4/aout.4
.\" Copyright (c) 2012 Konstantin Belousov <kib@FreeBSD.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $FreeBSD: releng/10.2/share/man/man4/aout.4 239306 2012-08-15 18:00:52Z kib $
.\"
.Dd August 14, 2012
.Dt AOUT 4
.Os
.Sh NAME
.Nm aout
.Nd kernel support for executing binary files in legacy a.out format
.Sh SYNOPSIS
.Bd -literal -offset indent
kldload a.out
.Ed
.Sh DESCRIPTION
The
.Xr a.out 5
executable format was used before the release of
.Fx 3.0 .
Since i386 was the only supported architecture at that time,
.Xr a.out 5
executables can only be activated on platforms that support
execution of i386 code, such as i386, amd64 and (partially) ia64.
.Pp
To add kernel support for old syscalls and old syscall invocation methods,
place the following options in the kernel configuration file:
.Bd -ragged -offset indent
.Cd "options COMPAT_43"
.br
.Cd "options COMPAT_FREEBSD32"
.Ed
.Pp
The
.Va COMPAT_FREEBSD32
option is only required on 64-bit CPU architectures.
.Pp
The
.Va aout.ko
module needs to be loaded with the
.Xr kldload 8
utility in order to support the
.Xr a.out 5
image activator:
.Bd -ragged -offset indent
.Ic kldload aout
.Ed
.Pp
Alternatively, to load the module at boot time, place the following line in
.Xr loader.conf 5 :
.Bd -literal -offset indent
aout_load="YES"
.Ed
.Pp
The
.Xr a.out 5
format was mainstream quite a long time ago.
Reasonable default settings and security requirements of
modern operating systems today contradict the default environment of that time
and require adjustments of the system to mimic natural environment for
old binaries.
.Pp
The following
.Xr sysctl 8
tunables are useful for this:
.Bl -tag -offset indent -width "XXXXXXXXXXXXXXXXXXXXXXXXX"
.It Xo Va security.bsd.map_at_zero
.Xc
Set to 1 to allow mapping of process pages at address 0.
Some very old
.Va ZMAGIC
executable images require text mapping at address 0.
.It Xo Va kern.pid_max
.Xc
Old versions of
.Fx
used signed 16-bit type for
.Vt pid_t .
Current kernels use 32-bit type for
.Vt pid_t ,
and allow process id's up to 99999.
Such values cannot be represented by old
.Vt pid_t ,
mostly causing issues for processes using
.Xr wait 2
syscalls, for example shells.
Set the sysctl to 30000 to work around the problem.
.It Xo Va kern.elf32.read_exec
.Xc
Set to 1 to force any accessible memory mapping performed by 32-bit
process to allow execution, see
.Xr mmap 2 .
Old i386 CPUs did not have a bit in PTE which disallowed execution
from the page, so many old programs did not specify
.Va PROT_EXEC
even for mapping of executable code.
The sysctl forces
.Va PROT_EXEC
if mapping has any access allowed at all.
The setting is only needed if the host architecture allows
non-executable mappings.
.El
.Sh SEE ALSO
.Xr execve 2 ,
.Xr a.out 5 ,
.Xr elf 5 ,
.Xr sysctl 8 .
.Sh HISTORY
The
.Xr a.out 5
executable format was used on ancient
.At
and served as the main executable format for
.Fx
from the beginning up to
.Fx 2.2.9 .
In
.Fx 3.0
it was superseded by
.Xr elf 5 .
.Sh AUTHORS
The
.Nm
manual page was written by
.An Konstantin Belousov Aq kib@FreeBSD.org .
.Sh BUGS
On 64bit architectures, not all wrappers for older syscalls are implemented.


usr.bin/gprof/aout.c
/*-
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
/* From: */
#ifndef lint
static char sccsid[] = "@(#)gprof.c	8.1 (Berkeley) 6/6/93";
#endif /* not lint */
#endif

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/10.2/usr.bin/gprof/aout.c 246783 2013-02-14 08:16:03Z charnier $");

#include <netinet/in.h>

#include <a.out.h>
#include <err.h>
#include <string.h>

#include "gprof.h"

static void getstrtab(FILE *, const char *);
static void getsymtab(FILE *, const char *);
static void gettextspace(FILE *);
static bool funcsymbol(struct nlist *);

static char	*strtab;		/* string table in core */
static long	ssiz;			/* size of the string table */
static struct	exec xbuf;		/* exec header of a.out */

/* Things which get -E excluded by default. */
static char	*excludes[] = { "mcount", "__mcleanup", NULL };

    /*
     * Set up string and symbol tables from a.out.
     *	and optionally the text space.
     * On return symbol table is sorted by value.
     *
     * Returns 0 on success, -1 on failure.
     */
int
aout_getnfile(const char *filename, char ***defaultEs)
{
    FILE	*nfile;

    nfile = fopen( filename ,"r");
    if (nfile == NULL)
	err( 1 , "%s", filename );
    fread(&xbuf, 1, sizeof(xbuf), nfile);
    if (N_BADMAG(xbuf)) {
	fclose(nfile);
	return -1;
    }
    getstrtab(nfile, filename);
    getsymtab(nfile, filename);
    gettextspace( nfile );
    fclose(nfile);
#   ifdef DEBUG
	if ( debug & AOUTDEBUG ) {
	    register int j;

	    for (j = 0; j < nname; j++){
		printf("[getnfile] 0X%08lx\t%s\n", nl[j].value, nl[j].name);
	    }
	}
#   endif /* DEBUG */
    *defaultEs = excludes;
    return 0;
}

static void
getstrtab(FILE *nfile, const char *filename)
{

    fseek(nfile, (long)(N_SYMOFF(xbuf) + xbuf.a_syms), 0);
    if (fread(&ssiz, sizeof (ssiz), 1, nfile) == 0)
	errx( 1 , "%s: no string table (old format?)" , filename );
    strtab = calloc(ssiz, 1);
    if (strtab == NULL)
	errx( 1 , "%s: no room for %ld bytes of string table", filename , ssiz);
    if (fread(strtab+sizeof(ssiz), ssiz-sizeof(ssiz), 1, nfile) != 1)
	errx( 1 , "%s: error reading string table" , filename );
}

    /*
     * Read in symbol table
     */
static void
getsymtab(FILE *nfile, const char *filename)
{
    register long	i;
    int			askfor;
    struct nlist	nbuf;

    /* pass1 - count symbols */
    fseek(nfile, (long)N_SYMOFF(xbuf), 0);
    nname = 0;
    for (i = xbuf.a_syms; i > 0; i -= sizeof(struct nlist)) {
	fread(&nbuf, sizeof(nbuf), 1, nfile);
	if ( ! funcsymbol( &nbuf ) ) {
	    continue;
	}
	nname++;
    }
    if (nname == 0)
	errx( 1 , "%s: no symbols" , filename );
    askfor = nname + 1;
    nl = (nltype *) calloc( askfor , sizeof(nltype) );
    if (nl == 0)
	errx( 1 , "no room for %zu bytes of symbol table" ,
		askfor * sizeof(nltype) );

    /* pass2 - read symbols */
    fseek(nfile, (long)N_SYMOFF(xbuf), 0);
    npe = nl;
    nname = 0;
    for (i = xbuf.a_syms; i > 0; i -= sizeof(struct nlist)) {
	fread(&nbuf, sizeof(nbuf), 1, nfile);
	if ( ! funcsymbol( &nbuf ) ) {
#	    ifdef DEBUG
		if ( debug & AOUTDEBUG ) {
		    printf( "[getsymtab] rejecting: 0x%x %s\n" ,
			    nbuf.n_type , strtab + nbuf.n_un.n_strx );
		}
#	    endif /* DEBUG */
	    continue;
	}
	npe->value = nbuf.n_value;
	npe->name = strtab+nbuf.n_un.n_strx;
#	ifdef DEBUG
	    if ( debug & AOUTDEBUG ) {
		printf( "[getsymtab] %d %s 0x%08lx\n" ,
			nname , npe -> name , npe -> value );
	    }
#	endif /* DEBUG */
	npe++;
	nname++;
    }
    npe->value = -1;
}

    /*
     *	read in the text space of an a.out file
     */
static void
gettextspace(FILE *nfile)
{

    textspace = (u_char *) malloc( xbuf.a_text );
    if ( textspace == 0 ) {
	warnx("no room for %u bytes of text space: can't do -c" ,
		  xbuf.a_text );
	return;
    }
    (void) fseek( nfile , N_TXTOFF( xbuf ) , 0 );
    if ( fread( textspace , 1 , xbuf.a_text , nfile ) != xbuf.a_text ) {
	warnx("couldn't read text space: can't do -c");
	free( textspace );
	textspace = 0;
	return;
    }
}

static bool
funcsymbol(struct nlist *nlistp)
{
    char	*name, c;

	/*
	 *	must be a text symbol,
	 *	and static text symbols don't qualify if aflag set.
	 */
    if ( ! (  ( nlistp -> n_type == ( N_TEXT | N_EXT ) )
	   || ( ( nlistp -> n_type == N_TEXT ) && ( aflag == 0 ) ) ) ) {
	return FALSE;
    }
	/*
	 *	name must start with an underscore if uflag is set.
	 *	can't have any `funny' characters in name,
	 *	where `funny' means `.' (.o file names)
	 *	need to make an exception for sparc .mul & co.
	 *	perhaps we should just drop this code entirely...
	 */
    name = strtab + nlistp -> n_un.n_strx;
    if ( uflag && *name != '_' )
	return FALSE;
#ifdef sparc
    if ( *name == '.' ) {
	char *p = name + 1;
	if ( *p == 'u' )
	    p++;
	if ( strcmp ( p, "mul" ) == 0 || strcmp ( p, "div" ) == 0 ||
	     strcmp ( p, "rem" ) == 0 )
		return TRUE;
    }
#endif
    while ( (c = *name++) ) {
	if ( c == '.' ) {
	    return FALSE;
	}
    }
    return TRUE;
}
